File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int[] nums;
    public TreeNode sortedArrayToBST(int[] nums) {
        this.nums = nums;
        return dfs(0, nums.length - 1);
    }
    
    TreeNode dfs(int l, int r) {
        if(l > r) return null;
        int m = l + (r - l) / 2;
        TreeNode root = new TreeNode(this.nums[m]);
        root.left = dfs(l, m - 1);
        root.right = dfs(m + 1, r);
        return root;
    }
}

File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int n = nums.size();
        function<TreeNode*(int, int)> dfs = [&](int left, int right) -> TreeNode*{
            if(left > right) return nullptr;
            int m = left + (right - left) / 2;
            TreeNode* root = new TreeNode(nums[m]);
            root->left = dfs(left, m - 1);
            root->right = dfs(m + 1, right);
            return root;
        };
        return dfs(0, n - 1);
    }
};